### PR TITLE
names labels

### LINKED ISSUE #220: add name to step when auto injecting reveal step

### LINKED ISSUE #227: Better default names
If the convention is going to be `U + Y + Z + Q + S + R + A` to build designs, we should build better default names for anonymous steps. eg

```
declare_population(N=100, foo=1) + declare_sampling(m=50)
```

should be named something like U and S instead of being named after the deparsed call. Perhaps it could be letter dot dot position.

For each step type, declare a default letter, and use that to make names that look like U..1, S..2, etc.

### PR DESCRIPTION
**Pull request by acoppock**
closes #220 
closes #227 
closes #246

### CODE REVIEW

File: R/construct_design.R [hunk 1]
```diff
@@ -82,41 +82,32 @@
...
     if (!inherits(rhs, "dd") && !inherits(rhs, "function")) {
       stop("The right hand side does not appear to be a DeclareDesign object or a function", call. = FALSE)
     }
     
     if (inherits(lhs, "design")) {
       
-     # auto_steps <- sapply(lhs, function(x) { lbl <- attr(x, "label"); !is.null(lbl) && grepl("Autogenerated by ", lbl) })
-     auto_steps <- sapply(lhs, function(x) !is_null(attr(x, "auto-generated")))
-        
+      auto_steps <- sapply(lhs, function(x) !is_null(attr(x, "auto-generated")))
```
Comment (nfultz) [inline R/construct_design.R]: pull out predicate to helper file, rewrite as Filter(Negate(is_autogenerated), lhs)

File: R/construct_design.R [hunk 2]
```diff
@@ -128,31 +119,19 @@
 construct_design <- function(...) {
   
   qs <- quos(...)
+  qs <- maybe_add_names_qs(qs)
```
Comment (nfultz) [inline R/construct_design.R]: note that you don't really use qnames now, can probably purge this and set names(ret) directly.

File: R/construct_design.R [hunk 3]
```diff
@@ -192,56 +187,6 @@ construct_design <- function(...) {
...
-  
   ret
 }
 
-###############################################################################
-# In a design created using the + operator, if a step is a dplyr style call mutate(foo = bar),
-# it will fail to evaluate - we can catch that and try to curry it
-
-#' @importFrom rlang quos lang_fn lang_modify eval_tidy
-callquos_to_step <- function(step_call, label="") {
```
Comment (nfultz) [inline R/construct_design.R:212]: +1

File: R/declare_potential_outcomes.R [hunk 4]
```diff
@@ -290,54 +290,32 @@ pofdv <- function(design, i, step){
...
-    new_step <- eval_tidy(quo(declare_reveal(outcome_variables = !!this_step_meta$outcome_variables,
-                                             assignment_variables = !!this_step_meta$assignment_variables,
-                                             label = !!paste("Autogenerated by", attr(step, "label"))) ))
-    attr(new_step, "auto-generated") <- TRUE
-    
-    for (step_j in design[length(design):(i + 1)]) {
-      if (is.null(step_meta <- attr(step_j, "step_meta"))) next;
-      if (attr(step_j, "step_type") == "assignment") {
-        if (any(step_meta$assignment_variables %in% attr(step, "step_meta")$assignment_variables)) {
-          # warning("Attempting to inject a `declare_reveal(", this_step_meta$outcome_variables, ", ",
```
Comment (nfultz) [inline R/declare_potential_outcomes.R:313]: The warnings should probably come back at some point.

File: R/utilities.R [hunk 5]
```diff
@@ -5,35 +5,53 @@
...
-    quotations[[i]] <- labeller(quotations[[i]], names(quotations)[i])
-  }
-  
-  if (any(duplicated(names(quotations)))) {
-    stop(paste0("Please provide unique names for each design step. Duplicates include ", 
-                paste(names(quotations)[duplicated(names(quotations))], collapse = ", "), 
-                ". You can name steps within declare design, e.g. declare_design(my_pop = pop_step)."), call. = FALSE)
+    if (names(quotations)[i] == ""){
+      names(quotations)[i] <- namer(quotations[[i]])
+    }
```
Comment (nfultz) [inline R/utilities.R:38]: whole loop can be vectorized now.

**Review by nfultz — CHANGES_REQUESTED** [review]

File: R/construct_design.R (hunk 1 above)
Comment (graemeblair) [reply R/construct_design.R]: thx. couldn't get this to work. tried: 

is_autogenerated <- function(steps) {
  sapply(steps, attr, "auto-generated")
}

lhs <- Filter(Negate(is_autogenerated), lhs)

am I missing something?

File: R/construct_design.R (hunk 2 above)
Comment (graemeblair) [reply R/construct_design.R]: we do that a bit lower, but need here to capture names like pop when sent to `pop + declare_sampling(n = 50)`. the name `pop` is lost if you don't use maybe_add_names on the quosures.

File: R/declare_potential_outcomes.R (hunk 4 above)
Comment (graemeblair) [reply R/declare_potential_outcomes.R:313]: thx, yes, will remember to come back to this

**Comment by nfultz** [comment]
See https://github.com/DeclareDesign/DeclareDesign/pull/249

File: R/construct_design.R (hunk 2 above)
Comment (nfultz) [reply R/construct_design.R]: There's only one place that calls in to conduct_design, should not need any NSE within it - pass the steps directly. See #249

File: tests/testthat/test-fanout.R [hunk 6]
```diff
@@ -84,6 +84,7 @@ test_that("sims expansion is correct",{
 
 
 test_that("fan_out ids are correct",{
+  sleep_large <- resample_data(sleep, N = 200)
```
Comment (nfultz) [inline tests/testthat/test-fanout.R]: is this used anywhere?

File: R/construct_design.R [hunk 7]
```diff
@@ -82,56 +82,56 @@
...
+  
+  for(i in match(to_rename, names(rhs))) {
+    if(attr(rhs, "valid")) next;
+    new_nm <- names(rhs)[i]
     
-    if (!inherits(rhs, "dd") && !inherits(rhs, "function")) {
-      stop("The right hand side does not appear to be a DeclareDesign object or a function", call. = FALSE)
-    }
+    while(new_nm %in% union(names(lhs), names(rhs)))
+      new_nm <- paste0(new_nm, "._")
```
Comment (nfultz) [inline R/construct_design.R]: Needs to add numbers like data.frame does.

**Review by nfultz — CHANGES_REQUESTED** [review]

File: tests/testthat/test-fanout.R (hunk 6 above)
Comment (graemeblair) [reply tests/testthat/test-fanout.R]: that test fails intermittently, was trying to see if upping sample size would help

File: R/construct_design.R (hunk 7 above)
Comment (graemeblair) [reply R/construct_design.R]: great, just made commits along these lines

**Comment by graemeblair** [comment]
think this is good to go, subject to @nfultz review

File: tests/testthat/test-fanout.R (hunk 6 above)
Comment (graemeblair) [reply tests/testthat/test-fanout.R]: oic I didn't actually implement. commit to follow

**Pull request merged**